internal/tofu/node_resource_abstract_instance.go

```diff
@@ -63,6 +68,44 @@ func NewNodeAbstractResourceInstance(addr addrs.AbsResourceInstance) *NodeAbstra
 	}
 }
 
+func (n *NodeAbstractResourceInstance) ProvidedBy() (map[addrs.InstanceKey]addrs.ProviderConfig, ExactProvider) {
+	return n.NodeAbstractResource.ProvidedByImpl(n.ResolvedInstanceProvider)
+}
+
+func (n *NodeAbstractResourceInstance) SetProvider(provider addrs.AbsProviderConfig, isResourceProvider bool) {
+	if isResourceProvider {
+		n.ResolvedResourceProvider = provider
+	} else {
+		n.ResolvedInstanceProvider = provider
+	}
+}
+
+func (n *NodeAbstractResourceInstance) ResolvedProvider() addrs.AbsProviderConfig {
+	var result addrs.AbsProviderConfig
+	if n.ResolvedResourceProvider.IsSet() {
+		result = n.ResolvedResourceProvider
+	} else {
+		// This function is called during the apply graph transformer phase to get the resolved provider.
+		// As this is an instance with a known address, we can lookup the resolved provider for the instance here..
+		if !n.ResolvedInstanceProvider.IsSet() {
+			n.ResolvedInstanceProvider = n.resolveInstanceProvider(n.Addr)
+		}
+
+		result = n.ResolvedInstanceProvider
+	}
+
+	if !result.IsSet() {
+		panic(fmt.Sprintf("ResolvedProvider for %s cannot get a provider", n.Addr))
+	}
+
+	// Throw an error if ResolvedResourceProvider and ResolvedInstanceProvider exists at the same time.
+	if n.ResolvedInstanceProvider.IsSet() && n.ResolvedResourceProvider.IsSet() {
+		panic(fmt.Sprintf("ResolvedProvider for %s has a provider set for the resource and the resource's instance", n.Addr))
+	}
```

Comment:
I am wondering if we can restructure code in such a way so those checks are compile-time guaranteed.

Reply:
Hmm, I can't think of a way to do it now, but there are parts I'm convinced we can refactor and want to do later after we are in alpha (like the provider transformer code, for example).